### PR TITLE
MAINT: Chain exceptions in several places.

### PR DESCRIPTION
**Pull request by keremh**
this solution is related to the following issue #15986
@eric-wieser can you review this pr?
Thank you.

### CODE REVIEW

File: numpy/lib/shape_base.py [hunk 1]
```diff
@@ -372,8 +372,8 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     # invoke the function on the first item
     try:
         ind0 = next(inds)
-    except StopIteration:
-        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
+    except StopIteration as e:
+        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0') from e
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py:376]:
```suggestion
    except StopIteration:
        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0') from None
```
`StopIteration` doesn't provide any useful context here

File: numpy/lib/index_tricks.py [hunk 2]
```diff
@@ -368,8 +368,8 @@ def __getitem__(self, key):
                         if len(vec) == 3:
                             trans1d = int(vec[2])
                         continue
-                    except Exception:
-                        raise ValueError("unknown special directive")
+                    except Exception as e:
+                        raise ValueError("unknown special directive") from e
```
Comment (eric-wieser) [inline numpy/lib/index_tricks.py:372]:
```suggestion
                    except Exception as e:
                        raise ValueError("unknown special directive {}".format(item)) from e
```

File: numpy/lib/shape_base.py (hunk 1 above)
Comment (keremh) [reply numpy/lib/shape_base.py]: i changed this line, thank you.

File: numpy/lib/index_tricks.py (hunk 2 above)
Comment (keremh) [reply numpy/lib/index_tricks.py]: updated this line too, thank you.

**Review by eric-wieser — APPROVED** [review]

File: numpy/lib/index_tricks.py [hunk 3]
```diff
@@ -368,8 +368,8 @@ def __getitem__(self, key):
                         if len(vec) == 3:
                             trans1d = int(vec[2])
                         continue
-                    except Exception:
-                        raise ValueError("unknown special directive")
+                    except Exception as e:
+                        raise ValueError("unknown special directive {!r}".format(item)) from e
```
Comment (rossbar) [inline numpy/lib/index_tricks.py:372]:
```suggestion
                        raise ValueError(
                            "unknown special directive {!r}".format(item)
                        ) from e
```
Style nit - 80 char limit

File: numpy/lib/shape_base.py [hunk 4]
```diff
@@ -372,8 +372,8 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     # invoke the function on the first item
     try:
         ind0 = next(inds)
-    except StopIteration:
-        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
+    except StopIteration as e:
+        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0') from None
```
Comment (rossbar) [inline numpy/lib/shape_base.py:376]:
```suggestion
        raise ValueError(
            'Cannot apply_along_axis when any iteration dimensions are 0'
        ) from None
```
Style nit - 80 char limit

**Review by rossbar — COMMENTED** [review]
A couple style nits re: line char limits

**Comment by keremh** [comment]
@rossbar thank you for your suggestions,
i changed commits.

**Comment by charris** [comment]
Thanks @keremh .

**Pull request merged**